File: web/packages/contracts/src/rest.ts
// REST contract — TS mirror of server/app/schemas/rest.py (§5.2).

export interface RegisterRequest {
  username: string;
  email: string;
  password: string;
}

export interface LoginRequest {
  username: string;
  password: string;
}

export interface TokenResponse {
  access_token: string;
  token_type: string;
}

export interface ClaimCardRequest {
  rfid: string;
  claim_code: string;
}

export interface NanosOut {
  nanos_type: string;
  level: number;
}

export interface MeResponse {
  id: number;
  username: string;
  level: number;
  xp: number;
  protein: number;
  carbs: number;
  lipids: number;
}

export interface InventoryResponse {
  protein: number;
  carbs: number;
  lipids: number;
  nanos: NanosOut[];
}

export interface UpgradeResponse {
  nanos_type: string;
  new_level: number;
  spent: Record<string, number>;
}

export interface FoodOut {
  id: number;
  food_class: string;
  display_name_zh: string | null;
  health: number;
  spoiled: boolean;
  claimed: boolean;
  reserved: boolean;
  reserved_by: number | null;
}

export interface NodeOut {
  id: string;
  name: string;
  location: string;
  status: string;
  last_heartbeat: string | null;
  lat: number | null;
  lng: number | null;
  tailscale_ip: string | null;
  entropy: number;
  health_avg: number;
}

export interface NodeDetail extends NodeOut {
  foods: FoodOut[];
}

export interface SalvageRequest {
  count: number;
}

export interface SalvageResponse {
  txn_id: string;
  gains: Record<string, number>;
  xp: number;
  claimed: number;
}

export interface BountyOut {
  id: number;
  spec_json: Record<string, unknown>;
  reward_json: Record<string, unknown>;
  progress: number;
  completed: boolean;
  claimed: boolean;
}

export interface FeedbackRequest {
  transaction_id: number;
  purity_stars: number;
}

export interface EsgStats {
  co2_saved_g: number;
  money_saved: number;
  meals_rescued: number;
}

export interface EntropyPoint {
  ts: string;
  total_entropy: number;
}

export interface EntropyStats {
  total_entropy: number;
  series: EntropyPoint[];
}

export interface FleetAccuracy {
  accuracy: number;
  sample_size: number;
}

export interface LeaderboardEntry {
  username: string;
  xp: number;
  level: number;
}

export interface TransactionOut {
  txn_id: string;
  node_id: string;
  items_json: Array<Record<string, unknown>>;
  gains_json: Record<string, unknown>;
  xp_awarded: number;
  created_at: string;
}

export interface Page<T = unknown> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

export interface ErrorResponse {
  detail: string;
}

File: server/app/schemas/rest.py
"""REST request/response contract — §5.2.

Response models mirror the DB shape but are decoupled from ORM internals.
Endpoints land in M4; these types are defined now so frontend + backend agree.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, EmailStr, Field

ORM = ConfigDict(from_attributes=True)


# ---- Auth / accounts ----
class RegisterRequest(BaseModel):
    username: str = Field(min_length=3, max_length=64)
    email: EmailStr
    password: str = Field(min_length=8)


class LoginRequest(BaseModel):
    username: str
    password: str


class TokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"


class ClaimCardRequest(BaseModel):
    rfid: str
    claim_code: str


# ---- Me / inventory ----
class NanosOut(BaseModel):
    model_config = ORM
    nanos_type: str
    level: int


class MeResponse(BaseModel):
    model_config = ORM
    id: int
    username: str
    level: int
    xp: int
    protein: float
    carbs: float
    lipids: float


class InventoryResponse(BaseModel):
    protein: float
    carbs: float
    lipids: float
    nanos: list[NanosOut]


class UpgradeResponse(BaseModel):
    nanos_type: str
    new_level: int
    spent: dict[str, float]


# ---- Nodes / foods ----
class FoodOut(BaseModel):
    model_config = ORM
    id: int
    food_class: str
    health: float
    spoiled: bool
    claimed: bool


class NodeOut(BaseModel):
    model_config = ORM
    id: str
    name: str
    location: str
    status: str
    last_heartbeat: datetime | None = None
    lat: float | None = None
    lng: float | None = None
    entropy: float = 0.0
    health_avg: float = 0.0


class NodeDetail(NodeOut):
    foods: list[FoodOut] = []


# ---- Bounties ----
class BountyOut(BaseModel):
    model_config = ORM
    id: int
    spec_json: dict[str, Any]
    reward_json: dict[str, Any]
    progress: float = 0.0
    completed: bool = False
    claimed: bool = False


# ---- Feedback ----
class FeedbackRequest(BaseModel):
    transaction_id: int
    purity_stars: int = Field(ge=1, le=5)


# ---- ESG / stats (UI2) ----
class EsgStats(BaseModel):
    co2_saved_g: float
    money_saved: float
    meals_rescued: int


class EntropyPoint(BaseModel):
    ts: datetime
    total_entropy: float


class EntropyStats(BaseModel):
    total_entropy: float
    series: list[EntropyPoint]


class FleetAccuracy(BaseModel):
    accuracy: float
    sample_size: int


class LeaderboardEntry(BaseModel):
    username: str
    xp: int
    level: int


class TransactionOut(BaseModel):
    model_config = ORM
    txn_id: str
    node_id: str
    items_json: list[dict[str, Any]]
    gains_json: dict[str, Any]
    xp_awarded: int
    created_at: datetime


class Page(BaseModel):
    items: list[Any]
    total: int
    page: int
    page_size: int


# ---- Generic error envelope ----
class ErrorResponse(BaseModel):
    detail: str
